### PR TITLE
[vec] support base64 function

### PR DESCRIPTION
**Pull request by zenoyang**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by HappenLee — COMMENTED** [review]
应该要考虑好返回null值得情况

**Comment by zenoyang** [comment]
Fixed the situation of returning null values, and added a util test @HappenLee

**Comment by HappenLee** [comment]
nice job！

File: be/src/vec/functions/function_string_to_string.h [hunk 1]
```diff
@@ -65,4 +66,51 @@ class FunctionStringToString : public IFunction {
     }
 };
 
+
+// func(string) -> nullable(type)
+template<typename Impl>
+class FunctionStringOperateToNullType : public IFunction {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string_to_string.h:72]: 如果是通用的to null type是不应该放到string_to_string这个头文件里的，你应该移动到function totype.h中

File: be/src/vec/functions/function_string_to_string.h [hunk 2]
```diff
@@ -65,4 +66,51 @@ class FunctionStringToString : public IFunction {
...
+    DataTypePtr get_return_type_impl(const DataTypes &arguments) const override {
+        return make_nullable(std::make_shared<typename Impl::ReturnType>());
+    }
+
+    bool use_default_implementation_for_constants() const override { return true; }
+
+    Status execute_impl(Block &block, const ColumnNumbers &arguments, size_t result,
+                        size_t input_rows_count) override {
+        auto null_map = ColumnUInt8::create(input_rows_count, 0);
+        ColumnPtr argument_columns[1];
```
Comment (HappenLee) [inline be/src/vec/functions/function_string_to_string.h:93]: 数组只有一个值，就别用数组了

File: be/src/vec/functions/function_string_to_string.h [hunk 3]
```diff
@@ -65,4 +66,51 @@ class FunctionStringToString : public IFunction {
...
+        argument_columns[0] =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
+        if (auto *nullable = check_and_get_column<ColumnNullable>(*argument_columns[0])) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string_to_string.h:96]: 需要处理nullable的话，需要设置use_default_implementation_for_nulls这个函数

File: be/src/vec/functions/function_string.cpp [hunk 4]
```diff
@@ -381,6 +383,86 @@ struct StringSpace {
...
+        auto rows_count = offsets.size();
+        dst_offsets.resize(rows_count);
+
+        for (int i = 0; i < rows_count; ++i) {
+            if (null_map[i]) {
+                StringOP::push_null_string(i, dst_data, dst_offsets, null_map);
+                continue;
+            }
+
+            const char* source = reinterpret_cast<const char*>(&data[offsets[i - 1]]);
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.cpp:402]: 右边有类型了，左边直接auto就行了

File: be/src/vec/functions/function_string.cpp [hunk 5]
```diff
@@ -381,6 +383,86 @@ struct StringSpace {
...
+            size_t srclen = offsets[i] - offsets[i - 1] - 1;
+
+            if (*source == '\0' && srclen == 0) {
+                StringOP::push_null_string(i, dst_data, dst_offsets, null_map);
+                continue;
+            }
+
+            int cipher_len = (int)(4.0 * ceil((double)srclen / 3.0));
+            boost::scoped_array<char> p;
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.cpp:411]: 这里一定要new吗？可否使用栈内存

File: be/src/vec/functions/function_string.cpp [hunk 6]
```diff
@@ -381,6 +383,86 @@ struct StringSpace {
...
+            const char* source = reinterpret_cast<const char*>(&data[offsets[i - 1]]);
+            size_t srclen = offsets[i] - offsets[i - 1] - 1;
+
+            if (*source == '\0' && srclen == 0) {
+                StringOP::push_null_string(i, dst_data, dst_offsets, null_map);
+                continue;
+            }
+
+            int cipher_len = srclen;
+            boost::scoped_array<char> p;
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.cpp:450]: 同上

File: be/test/vec/function/function_base64_test.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    SchemaScanner schema_scanner(column_descs, 1);
+    ObjectPool object_pool;
+    SchemaScannerParam param;
+    schema_scanner.init(&param, &object_pool);
+
+    auto tuple_desc = const_cast<TupleDescriptor*>(schema_scanner.tuple_desc());
+    RowDescriptor row_desc(tuple_desc, false);
+    auto tracker_ptr = MemTracker::CreateTracker(-1, "FunctionTest", nullptr, false);
+
+    int CAPACITY = 1024;
```
Comment (HappenLee) [inline be/test/vec/function/function_base64_test.cpp:69]: 同一个常量就别定义多次了

**Review by HappenLee — COMMENTED** [review]
有些小问题，请修订

**Comment by zenoyang** [comment]
> 有些小问题，请修订

以上问题已修复，有空再帮忙review，多谢~

**Pull request merged**